File: src-tauri/python-api/models/terminal.py
"""
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                        CEYBYTE POS                                               │
│                                                                                                  │
│                                     Terminal Model                                               │
│                                                                                                  │
│  Description: Terminal/workstation model for multi-terminal network support.                     │
│               Tracks terminal status, configuration, and sync information.                       │
│                                                                                                  │
│  Author: Akash Hasendra                                                                          │
│  Copyright: 2025 Ceybyte.com - Sri Lankan Point of Sale System                                   │
│  License: MIT License with Sri Lankan Business Terms                                             │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
"""

from sqlalchemy import Column, String, Integer, Boolean, Text, DateTime, JSON
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from database.base import BaseModel


class Terminal(BaseModel):
    """Terminal/workstation model for multi-terminal support"""
    
    __tablename__ = "terminals"
    
    # Terminal identification
    terminal_id = Column(String(50), unique=True, nullable=False, index=True)
    terminal_name = Column(String(100), nullable=False)
    display_name = Column(String(100), nullable=True)
    
    # Terminal type and role
    terminal_type = Column(String(20), default="pos")  # pos, admin, mobile
    is_main_terminal = Column(Boolean, default=False)
    
    # Network information
    ip_address = Column(String(45), nullable=True)  # IPv4 or IPv6
    mac_address = Column(String(17), nullable=True)
    hostname = Column(String(100), nullable=True)
    
    # Hardware information
    hardware_id = Column(String(100), nullable=True)  # Unique hardware identifier
    cpu_info = Column(String(200), nullable=True)
    memory_gb = Column(Integer, nullable=True)
    storage_gb = Column(Integer, nullable=True)
    
    # Software information
    os_version = Column(String(100), nullable=True)
    app_version = Column(String(20), nullable=True)
    last_update_check = Column(DateTime(timezone=True), nullable=True)
    
    # Terminal status
    status = Column(String(20), default="offline")  # online, offline, maintenance, error
    last_seen = Column(DateTime(timezone=True), nullable=True)
    last_heartbeat = Column(DateTime(timezone=True), nullable=True)
    
    # Sync information
    last_sync_time = Column(DateTime(timezone=True), nullable=True)
    sync_status = Column(String(20), default="pending")  # synced, pending, failed, conflict
    pending_sync_count = Column(Integer, default=0)
    
    # Configuration
    configuration = Column(JSON, nullable=True)  # Terminal-specific settings
    
    # Printer configuration
    printer_name = Column(String(100), nullable=True)
    printer_type = Column(String(50), nullable=True)  # thermal, inkjet, laser
    printer_width_mm = Column(Integer, default=80)
    
    # UPS information
    ups_connected = Column(Boolean, default=False)
    ups_battery_level = Column(Integer, nullable=True)  # 0-100
    ups_status = Column(String(20), nullable=True)  # online, battery, low_battery, critical
    
    # Performance metrics
    avg_response_time_ms = Column(Integer, nullable=True)
    error_count_24h = Column(Integer, default=0)
    
    # Location information
    location = Column(String(100), nullable=True)
    description = Column(Text, nullable=True)
    
    # Relationships
    sales = relationship("Sale", back_populates="terminal")
    
    def __repr__(self):
        return f"<Terminal(terminal_id='{self.terminal_id}', name='{self.terminal_name}', status='{self.status}')>"
    
    def is_online(self) -> bool:
        """Check if terminal is currently online"""
        return self.status == "online"
    
    def is_main(self) -> bool:
        """Check if this is the main terminal"""
        return self.is_main_terminal
    
    def needs_sync(self) -> bool:
        """Check if terminal needs synchronization"""
        return self.sync_status in ["pending", "failed"] or self.pending_sync_count > 0
    
    def update_heartbeat(self):
        """Update terminal heartbeat timestamp"""
        self.last_heartbeat = func.now()
        self.last_seen = func.now()
        if self.status == "offline":
            self.status = "online"
    
    def set_offline(self):
        """Mark terminal as offline"""
        self.status = "offline"
    
    def get_uptime_status(self) -> str:
        """Get terminal uptime status"""
        if not self.last_seen:
            return "Never connected"
        
        from datetime import datetime, timedelta
        now = datetime.now()
        time_diff = now - self.last_seen
        
        if time_diff < timedelta(minutes=5):
            return "Online"
        elif time_diff < timedelta(hours=1):
            return f"Last seen {time_diff.seconds // 60} minutes ago"
        elif time_diff < timedelta(days=1):
            return f"Last seen {time_diff.seconds // 3600} hours ago"
        else:
            return f"Last seen {time_diff.days} days ago"
    
    def get_sync_status_display(self) -> str:
        """Get user-friendly sync status"""
        status_names = {
            "synced": "Up to date",
            "pending": f"Pending ({self.pending_sync_count} items)",
            "failed": "Sync failed",
            "conflict": "Sync conflict"
        }
        return status_names.get(self.sync_status, self.sync_status.title())

File: src/types/terminal.ts
/**
 * ┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
 * │                                        CEYBYTE POS                                               │
 * │                                                                                                  │
 * │                                    Terminal Types                                                │
 * │                                                                                                  │
 * │  Description: TypeScript types for terminal management, multi-terminal networking,               │
 * │               and synchronization in the POS system.                                            │
 * │                                                                                                  │
 * │  Author: Akash Hasendra                                                                          │
 * │  Copyright: 2025 Ceybyte.com - Sri Lankan Point of Sale System                                   │
 * │  License: MIT License with Sri Lankan Business Terms                                             │
 * └──────────────────────────────────────────────────────────────────────────────────────────────────┘
 */

export type TerminalStatus = 'online' | 'offline' | 'maintenance' | 'error';
export type SyncStatus = 'synced' | 'pending' | 'failed' | 'conflict';
export type TerminalType = 'pos' | 'admin' | 'mobile';

export interface Terminal {
  terminal_id: string;
  terminal_name: string;
  display_name?: string;
  terminal_type: TerminalType;
  is_main_terminal: boolean;
  
  // Network information
  ip_address?: string;
  mac_address?: string;
  hostname?: string;
  
  // Hardware information
  hardware_id?: string;
  cpu_info?: string;
  memory_gb?: number;
  storage_gb?: number;
  
  // Software information
  os_version?: string;
  app_version?: string;
  last_update_check?: string;
  
  // Terminal status
  status: TerminalStatus;
  last_seen?: string;
  last_heartbeat?: string;
  
  // Sync information
  last_sync_time?: string;
  sync_status: SyncStatus;
  pending_sync_count: number;
  
  // Configuration
  configuration?: Record<string, any>;
  
  // Printer configuration
  printer_name?: string;
  printer_type?: string;
  printer_width_mm?: number;
  
  // UPS information
  ups_connected: boolean;
  ups_battery_level?: number;
  ups_status?: string;
  
  // Performance metrics
  avg_response_time_ms?: number;
  error_count_24h?: number;
  
  // Location information
  location?: string;
  description?: string;
  
  // Computed properties
  uptime_status: string;
  sync_status_display: string;
  is_online: boolean;
}

export interface TerminalConfig {
  terminal_id?: string;
  terminal_name: string;
  display_name?: string;
  terminal_type: TerminalType;
  is_main_terminal: boolean;
  app_version: string;
  network_path?: string;
}

export interface TerminalUpdate {
  terminal_name?: string;
  display_name?: string;
  status?: TerminalStatus;
  configuration?: Record<string, any>;
}

export interface SyncRequest {
  terminal_id: string;
  force_sync: boolean;
}

export interface SyncResult {
  success: boolean;
  synced_tables: string[];
  conflicts_resolved: number;
  errors: string[];
  sync_time: string;
  records_synced: number;
  message?: string;
}

export interface SyncStatusInfo {
  terminal_id: string;
  sync_status: SyncStatus;
  last_sync_time?: string;
  pending_sync_count: number;
  needs_sync: boolean;
  sync_status_display: string;
}

export interface NetworkConnectivity {
  network_available: boolean;
  main_computer_reachable: boolean;
  shared_folder_accessible: boolean;
  database_accessible: boolean;
  latency_ms?: number;
  error_message?: string;
}

export interface NetworkTestRequest {
  network_path?: string;
  terminal_type: string;
}

export interface NetworkDiagnostics {
  timestamp: string;
  network_connectivity: NetworkConnectivity;
  terminal_count: number;
  system_info: {
    platform: string;
    network_interfaces: string;
  };
}

export interface OfflineTransaction {
  id: string;
  type: 'sale' | 'customer_update' | 'product_update' | 'payment';
  data: Record<string, any>;
  queued_at: string;
  sync_status: 'pending' | 'synced' | 'failed';
  retry_count: number;
  last_retry?: string;
  error_message?: string;
}

export interface ConflictResolution {
  conflict_id: string;
  table_name: string;
  record_id: string;
  local_data: Record<string, any>;
  remote_data: Record<string, any>;
  local_timestamp: string;
  remote_timestamp: string;
  resolution_strategy: 'last_write_wins' | 'manual' | 'merge';
  resolved: boolean;
  winner?: 'local' | 'remote' | 'merged';
}

export interface TerminalHeartbeat {
  terminal_id: string;
  timestamp: string;
  status: TerminalStatus;
  performance_metrics?: {
    cpu_usage: number;
    memory_usage: number;
    disk_usage: number;
    response_time_ms: number;
  };
  ups_info?: {
    connected: boolean;
    battery_level?: number;
    status?: string;
  };
}

export interface TerminalDiscoveryResult {
  success: boolean;
  terminals: Terminal[];
  count: number;
  error?: string;
}

export interface TerminalApiResponse<T = any> {
  success: boolean;
  data?: T;
  message?: string;
  error?: string;
}

// API Response types
export interface InitializeTerminalResponse extends TerminalApiResponse {
  terminal_id: string;
}

export interface GetTerminalResponse extends TerminalApiResponse {
  terminal: Terminal;
}

export interface SyncTerminalResponse extends TerminalApiResponse {
  sync_result?: SyncResult;
}

export interface NetworkTestResponse extends TerminalApiResponse {
  connectivity: NetworkConnectivity;
}

export interface DiagnosticsResponse extends TerminalApiResponse {
  diagnostics: NetworkDiagnostics;
}

export interface OfflineDataResponse extends TerminalApiResponse {
  data: Record<string, any>[];
  count: number;
}

// Configuration constants
export const TERMINAL_STATUS_COLORS = {
  online: 'green',
  offline: 'red',
  maintenance: 'orange',
  error: 'red',
} as const;

export const SYNC_STATUS_COLORS = {
  synced: 'green',
  pending: 'orange',
  failed: 'red',
  conflict: 'purple',
} as const;

export const DEFAULT_HEARTBEAT_INTERVAL = 30000; // 30 seconds
export const DEFAULT_SYNC_INTERVAL = 60000; // 1 minute
export const MAX_RETRY_COUNT = 3;
export const OFFLINE_CACHE_EXPIRY = 24 * 60 * 60 * 1000; // 24 hours